src/batRandom.ml

```diff
@@ -73,6 +73,51 @@ let enum_char () = BatEnum.from char
 
 let choice e = BatEnum.drop (int (BatEnum.count e)) e; BatEnum.get_exn e
 
+let weighted_choice e =
+  let list, total_weight =
+    BatEnum.fold (fun (acc, total) ((_, w) as elem) ->
+        if w < 0. then
+          Printf.kprintf invalid_arg
+            "weighted_choice does not accept negative weights (here %f)"
+            w;
+        elem :: acc, total +. w) ([], 0.) e in
```

Comment:
This is what I'd do in 2017, because folds destroy readability:
```ocaml
let total = 0.0 in
e |> List.iter (fun ((_, w) as elem) ->
  if w < 0. then
    Printf.kprintf invalid_arg
      "weighted_choice does not accept negative weights (here %f)"
      w;
  total := !total +. w);
```

It also seems that reversing [e] into [list] is unnecessary.